python/orca/tutorial/NCF/pytorch_dataset.py

```diff
@@ -22,70 +22,143 @@
                 while (u, j) in self.train_mat:
                     j = np.random.randint(self.num_item)
-                self.features_ng.append([u, j])
-
-        labels_ps = [1 for _ in range(len(self.features_ps))]
-        labels_ng = [0 for _ in range(len(self.features_ng))]
-
-        self.features_fill = self.features_ps + self.features_ng
-        self.labels_fill = labels_ps + labels_ng
+                features_ng.append([u, j])
+        labels_ps = [1.0 for _ in range(len(features_ps))]
+        labels_ng = [0.0 for _ in range(len(features_ng))]
+        features_fill = features_ps + features_ng
+        labels_fill = labels_ps + labels_ng
+        self.data = pd.DataFrame(features_fill,
+                                 columns=["user", "item"], dtype=np.int32)
+        self.data['label'] = labels_fill
+
+    def train_test_split(self):
+        train_dataset, test_dataset = train_test_split(self.data, test_size=0.2, random_state=100)
+        return NCFData(train_dataset), NCFData(test_dataset)
+
+    def merge_features(self, users, items, total_cols):
+        self.data = users.merge(self.data, on='user')
+        self.data = self.data.merge(items, on='item')
+        self.data = self.data.loc[:, total_cols]
 
     def __len__(self):
-        return (self.num_ng + 1) * len(self.labels)
+        return len(self.data)
 
     def __getitem__(self, idx):
-        features = self.features_fill if self.is_sampling \
-            else self.features_ps
-        labels = self.labels_fill if self.is_sampling \
-            else self.labels
-
-        user = features[idx][0]
-        item = features[idx][1]
-        label = float(labels[idx])
-        return user, item, label
-
-
-def load_dataset(dataset_dir):
-    data_X = pd.read_csv(
+        return tuple(self.data.at[idx, i] for i in self.data.columns)
+
+
+def process_users_items(dataset_dir):
+    sparse_features = ['gender', 'zipcode', 'category']
+    dense_features = ['age']
+
+    users = pd.read_csv(
+        os.path.join(dataset_dir, 'users.dat'),
+        sep="::", header=None, names=['user', 'gender', 'age', 'occupation', 'zipcode'],
+        usecols=[0, 1, 2, 3, 4],
+        dtype={0: np.int32, 1: str, 2: np.int32, 3: np.int32, 4: str})
+    items = pd.read_csv(
+        os.path.join(dataset_dir, 'movies.dat'),
+        sep="::", header=None, names=['item', 'category'],
+        usecols=[0, 2], dtype={0: np.int32, 1: str})
+
+    user_num = users['user'].max() + 1
+    item_num = items['item'].max() + 1
+
+    # categorical encoding
+    for i in sparse_features:
+        df = users if i in users.columns else items
+        df[i], _ = pd.Series(df[i]).factorize()
+    sparse_features.append('occupation')  # occupation is already indexed.
+
+    # scale dense features
+    for i in dense_features:
+        scaler = MinMaxScaler()
+        df = users if i in users.columns else items
+        values = df[i].values.reshape(-1, 1)
```

Comment:
why need reshape here?

Reply:
Because MinMaxScaler needs the input to be 2-dim tensor, not 1-dim.

Reply:
Makes sense.